### PR TITLE
[Refactor:TAGrading] Migrate GradeableMessage to Vue

### PR DESCRIPTION
**Pull request by jndlansh**
### Why is this Change Important & Necessary?
In the ongoing migration of Submitty's UI from legacy Twig + jQuery to Vue.js, in this PR we migrate the `GradeableMessage `(Grader Responsibility Agreement), which a popup extended component only for Group 2,3&4 - Full access grader and  Limited access grader and peer grader. Relatively simple migration. Approx 300 lines of code for tests.

### What is the New Behavior?
Exacty same behavior as on main.

https://github.com/user-attachments/assets/cfefc61f-3c31-41f8-8ddb-1826154a62a4

### What steps should a reviewer take to reproduce or test the bug or new feature?

- Login as TA or Student
- Open gradeable that you can grade
- A Gradeable message should pop up on the first time
- Click on `Responsiblities as a Grader` button to explicitly see the message
- the message won't auto popup again is you press agree button on the popup window.

### Automated Testing & Documentation
- Cypress Component Testing added

https://github.com/user-attachments/assets/e7236c56-7ec1-4923-a4be-166b256a822e

### Other information
- Not a breaking change

### CODE REVIEW

**Review by GarvitKhandelwal31 — APPROVED** [review]
As for the preservation of the original functionality, everything looks the same to me. I tried verifying with different grading access roles, also after Accepting the pop-up does not come again. Nothing wrong in the code I could find.

Some tests are still failing, @jndlansh maybe you could verify if nothing is due to your PR changes, as I am unsure if so many of the tests are flaky.

**Comment by GarvitKhandelwal31** [comment]
image
One thing I wanted to ask, which might not be related to the PR is that  why are we having the close button. If the grader selects cancel then it redirects to the gradeable page not allowing to grade, which is fine. If he accepts he is allowed to grade.  But if the person presses the close button, he is allowed to  grade moreover the message keeps popping.
Tagging @bmcutler

**Review by roye2 — COMMENTED** [review]
The code changes look good and the functionality seems to work as intended. It looks like there are some cypress test failures that still need to be addressed, for example the TAGrading ones.

**Review by JManion32 — CHANGES_REQUESTED** [review]
Code changes are reasonable and functionality is completely preserved.

We just need to remove the bridge for `show-gradeable-message-review` in favor of the state management we discussed in our meeting today.

**Comment by jndlansh** [comment]
I believe this component is completely refactored in accordance with our plan of migration.

File: site/cypress/component/GradeableMessage.cy.js [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    describe('cancel action', () => {
+        it('emits cancel with courseUrl and hides popup', () => {
+            mountWithEmitSpy(GradeableMessage, 'cancel', baseProps);
+            cy.contains('button', 'Cancel').click();
+            cy.get('[data-testid="popup-window"]').should('not.exist');
+            cy.get('@eventHandler').should('have.been.calledOnce');
+        });
+    });
+
+    describe('edge cases', () => {
```
Comment (JManion32) [inline site/cypress/component/GradeableMessage.cy.js]: I don't think this edge case test is necessary. If there were special characters in a course or semester, there would be bigger problems to worry about than whether or not the Grader Responsibility Agreement popup loaded.

File: site/vue/src/components/GradeableMessage.vue [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    userGroup: number;
+    blindStatus: number;
+    courseUrl: string;
+}>();
+
+const emit = defineEmits<{
+    cancel: [url: string];
+}>();
+
+const storageKey = `${props.semester}-${props.course}-${props.gradeableId}-message`;
```
Comment (JManion32) [inline site/vue/src/components/GradeableMessage.vue]: Since we don't use `semester` and `gradeableId` elsewhere, why not just compute this in the PHP and send it in as a single prop?

The only other place we use `course` is in the emit, but we can just plug the `courseId` variable into the emit js like this:
`'cancel': '() => { window.location.href = courseId; }'`

This will let us pass 1 prop instead of 3.

**Review by JManion32 — CHANGES_REQUESTED** [review]
A few small things. Overall looks good.

File: site/app/templates/grading/electronic/Details.twig [hunk 3]
```diff
@@ -70,7 +70,21 @@
...
+            {% include 'Vue.twig' with {
+                'type': 'component',
+                'name': 'GradeableMessage',
+                'args': {
+                    'semester': semester,
+                    'course': course,
+                    'gradeableId': gradeable.getId,
+                    'userGroup': core.getUser().getGroup(),
+                    'blindStatus': blind_status,
+                    'courseUrl': course_url
```
Comment (JManion32) [inline site/app/templates/grading/electronic/Details.twig]: Remove the `courseUrl` prop, then change the event mapping to:
'cancel': '() => { window.location.href = course_url;

File: site/cypress/component/GradeableMessage.cy.js (hunk 1 above)
Comment (jndlansh) [reply site/cypress/component/GradeableMessage.cy.js]: Got it

File: site/app/templates/grading/electronic/Details.twig (hunk 3 above)
Comment (jndlansh) [reply site/app/templates/grading/electronic/Details.twig]: fixed

File: site/vue/src/components/GradeableMessage.vue (hunk 2 above)
Comment (jndlansh) [reply site/vue/src/components/GradeableMessage.vue]: fixed

**Review by JManion32 — COMMENTED** [review]
I was going to comment this inline but there are many instances of this:
`userGroup`'s value is the same as in `baseProps`. This is just a small quality control thing I've noticed on many of the component tests. Honestly, for this, there are only 3 props so it may be more readable to write them out each time.

Please also extend the popup component here. That will abstract all of that popup boilerplate, and probably allow you to remove a few of the component tests that the popup test already covers.

There is also a new button:

Main:
Image

Feature:
Image

Observed Functionality:
On your branch, `Cancel` closes the popup and redirects the user to the grading details page, while `Close` simply closes the popup.

On main, the `Cancel` button closes the popup and redirects the user to the grading details page.

**Comment by jndlansh** [comment]
I have fixed the component to use popup.vue + addressed all the other requested chnages. 

Also you mentioned in your comment -> "On main, the Cancel button closes the popup and redirects the user to the grading details page."

But actually it should take back to the Gradeables page and not the Details page, which is restored and preserved on the branch

**Comment by JManion32** [comment]
> I have fixed the component to use popup.vue + addressed all the other requested chnages.
> 
> Also you mentioned in your comment -> "On main, the Cancel button closes the popup and redirects the user to the grading details page."
> 
> But actually it should take back to the Gradeables page and not the Details page, which is restored and preserved on the branch

Whoops! You are correct.

**Review by JManion32 — CHANGES_REQUESTED** [review]
It looks like you aren't directly overriding the popup toggle button, which is adding a useless `Toggle popup (you should probably override this` button to the UI:
Image

Take a look at `Popup.vue` and other components that extend it to see how to override it. A decent example is `MarkSeenPopup.vue`.

**Comment by jndlansh** [comment]
Fixed the Popup (overriding the popup toggle button) issue. I am surprised I missed this. Probably it merged so well with the other buttons!

File: .commandcode/taste/taste.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# vue
```
Comment (JManion32) [inline .commandcode/taste/taste.md]: What is this? Just curious.

**Review by JManion32 — CHANGES_REQUESTED** [review]
I don't think you need to change `Popup.vue` here. It seems you are doing it so the `save` (primary) and `toggle` (cancel) buttons have custom functionality, but everything you need can be acheived via the Popup's props and emits. If `Popup.vue` didn't already support this, then I would've suggested adding it since this PR has the potential to cause inconsistent popup styling on the site.

I once again will point you to `MarkSeenPopup.vue`, which uses `Popup.vue`'s props and emits to customize the footer buttons.

**Comment by jndlansh** [comment]
Oh! I now understand the full power of Popup.vue
We have to develop the Dropdown component somewhat like this only 💪

**Review by JManion32 — APPROVED** [review]
Code and functionality lgtm. Tested multiple user types. Nice work.
